### PR TITLE
RSS Virtual Keyboard

### PR DESCRIPTION
**Pull request by aidfromdeagland**
1. task description: https://github.com/rolling-scopes-school/tasks/blob/master/tasks/codejam-virtual-keyboard.md
2. screenshot: 
изображение
3. demo: https://aidfromdeagland.github.io/virtual-keyboard/
4. prod date: 05. 04.2020 / deadline: 07.04.2020
5. Total
Task ( 110 / 110 )
--DOM generation via js (20/20);
--keypress of a real keyboard keys highlights the same keys for a virtual keyboard (10/10);
--language switch and current language store via localStorage (15/15);
--mouse clicks and keyboard presses put an appropriate symbol to input (15/15);
--keypress animation (15/15);
--usage of ES6 features (15/15);
--eslint usage (10/10);
--following the PR/commits/repo rules (10/10);

PS. I have an unexpected behavior of the keyup event for 'ctrl' keys, while 'shift' key hold. For every single 'ctrl' keyup while 'shift' is pressed I, have next behavior: 
'ctrl' keyup - 2 keyup events fired,
'ctrl' keyup - 1 keyup event fired,
'ctrl' keyup - 2 keyup events fired,
'ctrl' keyup - 1 keyup event fired, and so on... 2-1-2-1-2-1-2-1-2-1...
I don't think, that the problem is in my code because I have the same trick at the test stand:  https://learn.javascript.ru/keyboard-events#keyboard-test-stand .

So I decided to change language switch key combo to 'alt' + 'shift'.  This works well.

mark calculation: 110 - 0 = 110.

### CODE REVIEW

File: src/keyboard.js [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    wrapper.appendChild(keyboard);
+    keyboard.appendChild(this.addButtons());
+    wrapper.appendChild(info);
+    document.body.appendChild(wrapper);
+    const shiftKeys = document.querySelectorAll('[data-code*="Shift');
+    const capslockKey = document.querySelector('[data-code="CapsLock"');
+    capslockKey.classList.add('keyboard__button_capslock');
+
+    document.addEventListener('keydown', (evt) => {
+      if (Object.keys(buttonsMap).includes(evt.code)) {
```
Comment (oltarium) [inline src/keyboard.js:66]: possible it can be replaced to 
```
if (buttonsMap[evt.code]) {}
```

File: src/keyboard.js [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        evt.preventDefault();
+        if ((evt.code === 'ShiftLeft' || evt.code === 'ShiftRight')) {
+          if (!Array.from(shiftKeys).some((x) => x.classList.contains('keyboard__button_active'))) {
```
Comment (oltarium) [inline src/keyboard.js:69]: please replace X to more meaningful variable name

File: src/keyboard.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
+import buttonsMap from './buttonsmap.js';
+
+class Keyboard {
+  constructor() {
+    this.lang = localStorage.getItem('keyboardLang') || 'en';
```
[the remark this answers is not in the document]
Comment (aidfromdeagland) [reply src/keyboard.js]: fixed

File: src/keyboard.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    this.capitalisation = 'normal';
```
[the remark this answers is not in the document]
Comment (aidfromdeagland) [reply src/keyboard.js]: fixed

File: src/keyboard.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+      button.classList.add(`keyboard__button_width_${buttonsMap[x].width}`);
+      button.dataset.code = x;
+      fragment.appendChild(button);
+    });
+
+    return fragment;
+  }
+
+  switchLanguage() {
+    this.lang = this.lang === 'en' ? 'ru' : 'en';
```
[the remark this answers is not in the document]
Comment (aidfromdeagland) [reply src/keyboard.js]: done

File: index.js [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import Keyboard from './src/keyboard.js';
+
+window.addEventListener('load', () => {
```
[the remark this answers is not in the document]
Comment (aidfromdeagland) [reply index.js]: I agree it's not necessary if my script is located at the end of the <body>.
fixed.

File: index.html [hunk 7]
```diff
@@ -4,10 +4,11 @@
   <meta charset="utf-8">
   <meta content="width=device-width, initial-scale=1" name="viewport">
   <title>Virtual keyboard RSSchool 2020q1</title>
+  <link as="script" href="index.js" rel="preload">
```
[the remark this answers is not in the document]
Comment (aidfromdeagland) [reply index.html]: I thought this sneaky move let me win a few nanoseconds for better loading time :D
fixed.

File: src/keyboard.js (hunk 1 above)
Comment (aidfromdeagland) [reply src/keyboard.js]: Yea. Clear and productive.
fixed.

File: src/keyboard.js (hunk 2 above)
Comment (aidfromdeagland) [reply src/keyboard.js]: done.

**Comment by aidfromdeagland** [comment]
> 
> 
> there is a bug. select two letters, and then pless any key on keyboard. The carriage sets before inserted letter.
> Initial score: 109

Thank you for your attentiveness. During fixing this bug I also make TAB and ENTER behave more friendly for user.

Thanks for feedback!

**Pull request merged**